### ISSUE TITLE
[Feature Request] Implement iLink capabilities

### ISSUE THREAD

**Issue by gtgamer468** [PCSX2/pcsx2#4472]
**Description**
 I would like Firewire and iLink to be fully implemented.

**Reason**
 The Firewire has always been a stub and never was implemented so to me, this seems long overdue. 

 The iLink feature was used in over a dozen titles, mostly first party ones such as Gran Turismo 3 and Concept and providing support for it would allow such games to be played like LAN titles paired with a VPN like Radmin.
 There's currently nothing in terms of support as iLink hasn't been implemented with any plugins in the past and the plugin system is going to be fully gone soon.

**Comment by RedDevilus** [comment]
Could maybe need a new label like Firewire?

**Comment by darkphoenixfox** [comment]
iLink support would be great to be able to play TIme Crisis 2 and Time Crisis 3 with 2 separate screens and 2 modern lightguns like the Sinden

**Comment by Thigoron** [comment]
If this would be implemented I'd be making a twin arcade cabinet. ^^

**Comment by shanethmoore** [comment]
I actually thought this already existed in the wild https://youtu.be/-K6rgnaZoms

Was very surprised to see its not.

Does anyone have any ballpark of how difficult this will be to develop?

**Comment by refractionpcsx2** [comment]
I call bullshit on that video.  Light guns don't work on LCD screens, and no, there's no iLink in PCSX2 right now.

**Comment by Thigoron** [comment]
> 

Sinden lightgun's are not "true" light guns. They do work on LCD's. Not sure how he ran the games in iLink tho'.

**Comment by Thigoron** [comment]
> I call bullshit on that video. Light guns don't work on LCD screens, and no, there's no iLink in PCSX2 right now.


Sinden lightgun's are not "true" light guns. They do work on LCD's. Not sure how he ran the games in iLink tho'.

**Comment by refractionpcsx2** [comment]
oh they're custom light guns? okay.  Yeah I have no idea, we've never implemented ilink.

**Comment by shanethmoore** [comment]
@refractionpcsx2 I have 2 of the sinden guns with the recoil, gotta say they're pretty awesome. 

Its basically just a HID mouse with a webcam attached. Theres No IR sensors or CRT needed.

They come with software that paints a white bezel around the game window. Then theres a camera build into the gun which uses some trained model to detect where the sqaure is painted and then from that it determines where the mouse cursor should go.


That all said, I have no idea how the guy in that video managed to do what he's done. I've seen examples where people literally drag their pcsx2 instance halfway across two screens and then use a white border on each, but I don't think thats what he's doing here as he somehow managed to get P1's screen centered right in the middle of Monitor 0. It looks like he did some split screen magic.

Anywho, I was just asking if there were any sizes on how complex ilink would be to implement. If I recall it was some weird firewire cable that had a 6mbs limit. I expect that would be hard to replicate across two pcsx2 instances but I'm hoping someone will prove me wrong.

**Comment by darkphoenixfox** [comment]
> I call bullshit on that video. Light guns don't work on LCD screens, and no, there's no iLink in PCSX2 right now.

That is _my_ video.

There is no iLink on PCSX2. That video was achieved by playing the single-console 2 Player mode and stretching the picture to fill two screens. Calibration for the guns didn't really work.

P1 is centered on the ultrawide screen because the screen was set up in Windows to a lower non-ultrawide resolution, but it's the trick @shanethmoore described.

[Sindens](https://www.sindenlightgun.com/) work on [any screen](https://www.youtube.com/watch?v=wZ8JQNxExNM) as stated above. 

I wish we didn't need to resort to alchemy-level tricks to get something like this working.

**Comment by refractionpcsx2** [comment]
That would explain it

**Comment by shanethmoore** [comment]
@darkphoenixfox Good to meet you man 👋  

Aye, calibration for TC2 has been a pain with Sindens. I'm doing some autoloading of state once the calibration is complete just to avoid going through it over and over. Tempted to try just using lilipad with a mouse instead of nuvee to see how it works.

You had comments disabled on YT so I wasn't able to ask how you did it. Indeed that is some top tier alchemy ☣️  
So your left hand screen has a narrower resolution than the monitor supports and thats how you got the 4:3 centered bezel. Clever. And your pcsx2 instance is probably windowed (as oppossed to maximized) and the application window header is hidden behind the sinden border.

Makes total sense now (if I've got that right)

**Comment by darkphoenixfox** [comment]
@shanethmoore you are correct .
Don't want to derrail the thread; if you want to talk more about sindens,etc ask for JCT in the Sinden discord

**Comment by shanethmoore** [comment]
Agreed!

So back on track, I still think Ilink support would be neat. Whats the best way for muggles and/or devs to support this?

**Comment by RedDevilus** [comment]
> Agreed!
> 
> So back on track, I still think Ilink support would be neat. Whats the best way for muggles and/or devs to support this?

Foremost there has to be some interest in it, the knowledge and lastly the research in the protocols and how it differs from the normal FireWire if at all.

It's pretty niche all things considered and not many games support, and you can see nobody bothered making an iLink plugin besides null, but that's just a placeholder without any real functionality.

I can understand you making that simple inquiry but there isn't an easy and concise way to answer to finally get the support.

**Comment by shanethmoore** [comment]
Good to have a todo list, thanks @RedDevilus

**Comment by ProfgLX** [comment]
Hi, I'm an admin from the sinden discord. iLink (2 players time crisis 2 and 3) and lightgun support in pcsx2 are some of the most asked about subjects in the discord, which has 7400 members. Same for the users of gun4ir, another lightgun system. Niche yes, but very popular still. Love your work btw.

**Comment by titchgamer** [comment]
Another Sinden/Gun4IR user here and can confirm what Prof says to be true.
Lightgun support for PCSX2 Is discussed daily in the discords along with 2 player options.

Time Crisis is a beloved franchise and one of the most popular requested games to be played.

if you can get easier gun support along with Ilink support you will make 1000’s of people very happy!!

**Comment by shanethmoore** [comment]
I mean I'd be happy to hit that sponsor button if I knew iLink was on the menu ;)

(I wonder if other discordians would feel the same. Only slightly kidding... I'll be more liquid in 2 months and can start contributing then anyways)

**Comment by stenzek** [comment]
Nagging for it and saying how things like how many people are waiting isn't going to make it happen any faster.

The USB subsystem is a complete mess and needs a fairly large refactor before it can be added to Qt, let alone adding lightgun support. Personally it doesn't really interest me, so I'm not doing it, however anyone else is welcome to, and it would be appreciated.

But until someone takes it on, please stop effectively spamming the thread, it's not helpful at all.

**Comment by shanethmoore** [comment]
I think those people were just trying to make the case that theres more demand out there as initially thought rather than spamming or nagging. If theres more tangible things people can help with outside of sponsoring then that might be a more productive way to make this feature deliverable

I'm just trying to see whats required to make it happen. So far I've listed

- [ ] USB subsystem refactor
- [ ] Some sort of abstraction for interaction with the protocols for iLink
- [ ] Probably needs to come after 1.7 because the plugin system is going away

I think Lightgun support is probably out of scope. If mice work in 1.7+ then I imagine most HID compliant devices would too.

@stenzek Fair enough if it doesn't interest you. I'd love to take it on but I'm coming in blind. If you've any more comprehensive or high level lists TODO's to make this happen, please do let us know.

**Comment by refractionpcsx2** [comment]
It's worth noting that we *are* an open source project.  If somebody else wants to implement it, it's welcomed and very possible, you don't have to wait for us to do it.  Of course we will check the code when it's submitted, and we frown upon people just stealing other peoples code and committing it (This happens a lot more often than I would like), so please don't do that.

**Comment by RageAShadey** [comment]
*A wild Snake appears*
Seems Like a good test of my skills. I'm on it >:3

**Comment by jamesbailey999** [comment]
Hi all,

I just wanted to resurrect this and to see if this is still being worked on as mentioned above?

Thanks for any update someone may give

**Comment by stenzek** [comment]
Nothing has changed (nobody working on it, no plans), please don't bump/spam threads - it doesn't achieve anything.

**Comment by Florin9doi** [comment]
https://web.archive.org/web/*/http://www.lsilogic.com:80/files/docs/techdocs/interface_cores/1394/*

**Comment by StillPlay2** [comment]
I dumped my iLink ID from my own SCPH-39003 to force it to read on PCSX2 to get around the issue but spews errors on system console

**Comment by ClonkAndre** [comment]
I'm all for iLink support!

**Comment by cuttherug** [comment]
Just here to give my plus 1 for iLink over LAN if possible. Thanks devs, and contributors!

**Comment by Florin9doi** [comment]
The cards listed as compatible by the nosy [documentation](https://github.com/torvalds/linux/blob/v7.0/drivers/firewire/Kconfig#L150-L154) are unobtainable, however, there are some lesser-known ones. I just got one, and in case anyone else is interested, I have a list of PCI Lynx cards: https://gist.github.com/Florin9doi/f1c4da7072f54243231629bf1bb7dd63

Some of them are currently available for cheap, especially when incorrect terms are used, like Solectron->Selectron or 2330->2230.

**Comment by sketchioh-hub** [comment]
iLink, (real!?)

**Comment by Florin9doi** [comment]
Patch to enable iLink logs for Gran Turismo SDS. Still not bootable:
```diff
 PCPX-96634:
   name: "GRAN TURISMO - SUBARU ドライビングシミュレーターVer."
   name-sort: "ぐらんつーりすも すばるどらいびんぐしみゅれーたーばーじょん"
   name-en: "Gran Tourismo - SUBARU Driving Simulator ver."
+  patches:
+    default:
+      content: |-
+        comment=sce1394Debug : 0x1001=1, 0x1002=0
+        patch=1,IOP,000addcc,word,00000000
+        patch=1,IOP,000addd0,word,00000001
```

**Comment by StillPlay2** [comment]
Battle Gear 2 utilises iLink to link up for multiplayer
